File: sd_modern_fortran/src/droplet.f90
!> Droplet data structure and operations module
!!
!! This module defines the core droplet_t type and associated operations
!! for the superdroplet method. Each superdroplet represents a collection
!! of identical real droplets.
!!
!! @author Daniel Rothenberg
!! @date November 2025
module droplet_mod

    use constants, only: dp, FOUR_THIRD, PI, RHO_WATER, THIRD
    implicit none
    private

    public :: droplet_t  ! droplet type
    public :: droplet_t_ ! droplet constructor bound to init_droplet
    public :: total_water
    public :: set_droplet

    !> Superdroplet data structure
    !!
    !! Represents a collection of identical real droplets in the superdroplet method.
    !! Contains both physical properties and derived quantities.
    !!
    !! @note Terminal velocity is cached and only recomputed when rcubed changes
    !!       for performance optimization.
    type :: droplet_t
        integer :: multi                !< Multiplicity (number of real droplets represented)
        real(kind=dp) :: &
            rcubed,                     & !< Cube of droplet radius (m^3)
            solute,                     & !< Solute mass (kg)
            density,                    & !< Droplet density (kg/m^3)
            radius,                     & !< Droplet radius (m)
            volume,                     & !< Droplet volume (m^3)
            mass,                       & !< Single droplet mass (kg)
            terminal_velocity             !< Terminal fall velocity (m/s)
    contains
        procedure, public :: set_droplet  !< Update droplet properties
    end type droplet_t

    interface droplet_t_
        module procedure init_droplet
    end interface

contains

    !> Constructor for droplet type
    !!
    !! Initializes a droplet with given multiplicity and radius.
    !! Computes all derived properties including volume, mass, and terminal velocity.
    !!
    !! @param[in] multi - Droplet multiplicity (number of real droplets represented)
    !! @param[in] radius - Droplet radius in meters
    !! @return Initialized droplet_t instance
    type(droplet_t) function init_droplet(multi, radius)

        integer, intent(in) :: multi
        real(kind=dp), intent(in) :: radius
        
        init_droplet%multi = multi
        init_droplet%radius = radius
        init_droplet%rcubed = radius**3.0_dp
        init_droplet%solute = 0.0_dp
        init_droplet%density = RHO_WATER
        init_droplet%volume = (4.0_dp / 3.0_dp) * PI * radius**3.0_dp
        init_droplet%mass = init_droplet%volume * init_droplet%density
        init_droplet%terminal_velocity = compute_terminal_velocity(radius, init_droplet%mass)

    end function init_droplet

    !> Mutator to update droplet properties
    !!
    !! Updates droplet multiplicity and rcubed value, recomputing derived
    !! properties only when rcubed has actually changed.
    !!
    !! @note Terminal velocity is only recomputed
    !!       when rcubed changes to avoid expensive calculations.
    !!
    !! @param[inout] self - Droplet instance to update
    !! @param[in] multi - New multiplicity value
    !! @param[in] rcubed - New radius cubed value (m^3)
    !! @param[in] solute - Optional new solute mass (kg)
    pure subroutine set_droplet(self, multi, rcubed, solute)

        class(droplet_t), intent(inout) :: self
        integer, intent(in) :: multi
        real(kind=dp), intent(in) :: rcubed
        real(kind=dp), intent(in), optional :: solute
        logical :: rcubed_changed

        ! Check if rcubed actually changed (avoid expensive terminal velocity recomputation)
        rcubed_changed = (abs(self%rcubed - rcubed) > 1e-20_dp)

        self%multi = multi
        self%rcubed = rcubed

        ! Only recompute derived properties if rcubed changed
        if (rcubed_changed) then
            self%radius = rcubed**(THIRD)
            self%volume = FOUR_THIRD * PI * rcubed  ! Optimized: use rcubed directly
            self%mass = self%volume * self%density
            self%terminal_velocity = compute_terminal_velocity(self%radius, self%mass)
        end if

        if ( present(solute) ) self%solute = solute

    end subroutine set_droplet

    !> Compute droplet terminal fall velocity
    !!
    !! Calculates terminal velocity using size-dependent power-law relationships
    !! based on Beard (1976) parameterization.
    !!
    !! @param[in] radius - Droplet radius (m)
    !! @param[in] mass - Droplet mass (kg)
    !! @return Terminal velocity in m/s
    pure function compute_terminal_velocity(radius, mass) result(tv)
        real(kind=dp), intent(in) :: radius, mass
        real(kind=dp) :: tv

        real(kind=dp) :: d, cbrt_x, alpha, x_to_beta

        d = 2.0_dp * radius * 1e6_dp ! diameter, m -> μm
        cbrt_x = (mass * 1e3_dp) ** THIRD ! cbr(mass), [kg -> g]^(1/3)

        if (d <= 134.43_dp) then 
            alpha = 4.5795e5_dp
            x_to_beta = cbrt_x * cbrt_x
        else if ((134.34_dp < d) .and. (d <= 1511.64_dp)) then
            alpha = 4962.0_dp
            x_to_beta = cbrt_x
        else if ((1511.64_dp < d) .and. (d <= 3477.84)) then
            alpha = 1732.0_dp
            x_to_beta = sqrt(cbrt_x)
        else
            alpha = 917.0_dp
            x_to_beta = 1.0_dp
        end if

        tv = 1e-2_dp * alpha * x_to_beta ! cm/s -> m/s

    end function compute_terminal_velocity

    !> Calculate total water mass across all superdroplets
    !!
    !! Sums the total water mass by multiplying each droplet's mass by its
    !! multiplicity and summing over all superdroplets.
    !!
    !! @param[in] droplets - Array of superdroplets
    !! @return Total water mass in kg
    pure function total_water(droplets) result(total_mass)
        type(droplet_t), dimension(:), intent(in) :: droplets
        real(kind=dp) :: total_mass
        integer :: i

        total_mass = 0.0_dp
        do i = 1, size(droplets)
            total_mass = total_mass + droplets(i)%mass * real(droplets(i)%multi, kind=dp)
        end do

    end function total_water

end module droplet_mod

File: sd_numba/src/sd_numba/core/droplet.py
"""
Superdroplet data structure and related methods.
"""

import numpy as np
from numba import jit, types
from numba.experimental import jitclass

from sd_numba.core.constants import FOUR_THIRD, PI, RHO_WATER


@jitclass(
    spec=[
        ("multi", types.int64),
        ("rcubed", types.float64),
        ("solute", types.float64),
        ("density", types.float64),
        ("radius", types.float64),
        ("volume", types.float64),
        ("mass", types.float64),
        ("terminal_velocity", types.float64),
    ]
)
class Droplet:
    """Represents a superdroplet in the simulation.

    A superdroplet is a computational particle that represents many
    identical physical droplets. The `multi` field indicates how many
    real droplets this superdroplet represents.

    In this implementation, we eagerly compute and cache important derived properties
    to avoid expensive recalculations.

    Attributes:
        multi (int): The multiplicity of the superdroplet.
        rcubed (float): The cube of the droplet radius (m^3).
        solute (float): The solute mass of the droplet (kg).
        density (float): The density of the droplet (kg/m^3).
        radius (float): The radius of the droplet (m).
        volume (float): The volume of the droplet (m^3).
        mass (float): The mass of the droplet (kg).
        terminal_velocity (float): The terminal velocity of the droplet (m/s).
    """

    multi: int
    rcubed: float
    solute: float
    density: float
    radius: float
    volume: float
    mass: float
    terminal_velocity: float

    def __init__(
        self,
        multi: int,
        rcubed: float,
        solute: float,
        density: float,
        radius: float,
        volume: float,
        mass: float,
        terminal_velocity: float,
    ):
        self.multi = multi
        self.rcubed = rcubed
        self.solute = solute
        self.density = density
        self.radius = radius
        self.volume = volume
        self.mass = mass
        self.terminal_velocity = terminal_velocity

    @staticmethod
    def new(multi: int, radius: float) -> "Droplet":
        """Create a new Droplet instance.

        Args:
            multi (int): The multiplicity of the superdroplet.
            radius (float): The radius of the droplet in meters.

        Returns:
            Droplet: A new Droplet instance.
        """
        rcubed = radius * radius * radius
        volume = rcubed * FOUR_THIRD * PI
        mass = volume * RHO_WATER
        terminal_velocity = compute_terminal_velocity(radius, mass)

        return Droplet(
            multi,
            rcubed,
            0.0,
            RHO_WATER,
            radius,
            volume,
            mass,
            terminal_velocity,
        )

    def update_rcubed(self, new_rcubed: float):
        """Update the rcubed value and recalculate the derived properties."""
        self.rcubed = new_rcubed
        self.radius = np.cbrt(new_rcubed)
        self.volume = new_rcubed * FOUR_THIRD * PI
        self.mass = self.volume * RHO_WATER
        self.terminal_velocity = compute_terminal_velocity(self.radius, self.mass)


@jit(nopython=True, fastmath=True)
def compute_terminal_velocity(radius: float, mass: float) -> float:
    """Compute the terminal velocity of a droplet using the Beard (1976) parameterization.

    Args:
        radius: The radius of the droplet in meters.
        mass: The mass of the droplet in kilograms.

    Returns:
        The terminal velocity of the droplet in meters per second.
    """
    d = 2.0 * radius * 1e6  # diameter, m -> μm
    x = mass * 1e3  # mass, kg -> g

    if d <= 134.43:
        alpha = 4.5795e5
        cbrt_x = np.cbrt(x)
        # x**(2/3) = cbrt(x)^2 = cbrt(x) * cbrt(x)
        x_to_beta = cbrt_x * cbrt_x
    elif d <= 1511.64:
        alpha = 4962.0
        x_to_beta = np.cbrt(x)
    elif d <= 3477.84:
        alpha = 1732.0
        # x**(1/6) = sqrt(cbrt(x))
        x_to_beta = np.sqrt(np.cbrt(x))
    else:
        alpha = 917.0
        x_to_beta = 1.0

    return 1e-2 * alpha * x_to_beta  # cm/s -> m/s


@jit(nopython=True, fastmath=True)
def _compute_total_water_jitted(droplets) -> float:
    """Compute the total water mass using jitted loop.

    Args:
        droplets: TypedList of Droplet instances

    Returns:
        The total water mass in kilograms.
    """
    total = 0.0
    for droplet in droplets:
        total += droplet.mass * droplet.multi
    return total


def compute_total_water(droplets) -> float:
    """Compute the total water mass of a list of droplets.

    Args:
        droplets: The list of Droplet instances (TypedList or list)

    Returns:
        The total water mass in kilograms.
    """
    return _compute_total_water_jitted(droplets)
